File: src-tauri/src/domain/money.rs
//! Integer-safe monetary helpers (minor units / cents).

/// Quantity scale: 6 decimal places (micro-units).
const QTY_SCALE: i128 = 1_000_000;

/// Line total in minor units from quantity × unit price.
///
/// Quantity is quantized to 6 decimal places, then multiplied in integer
/// arithmetic and rounded half-away-from-zero. Negative or non-finite results
/// clamp to 0 (same contract as the previous f64 helper).
pub fn line_total_minor(qty: f64, unit_minor: i64) -> i64 {
    if !qty.is_finite() {
        return 0;
    }
    let qty_scaled = (qty * QTY_SCALE as f64).round() as i128;
    let product = qty_scaled.saturating_mul(i128::from(unit_minor));
    let half = QTY_SCALE / 2;
    let rounded = if product >= 0 {
        (product + half) / QTY_SCALE
    } else {
        (product - half) / QTY_SCALE
    };
    if rounded <= 0 {
        0
    } else if rounded > i128::from(i64::MAX) {
        i64::MAX
    } else {
        rounded as i64
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn whole_quantity_exact() {
        assert_eq!(line_total_minor(2.0, 1250), 2500);
    }

    #[test]
    fn fractional_quantity_half_up() {
        // 0.1 × 33 → 3.3 → 3
        assert_eq!(line_total_minor(0.1, 33), 3);
        // 1.5 × 100 → 150
        assert_eq!(line_total_minor(1.5, 100), 150);
    }

    #[test]
    fn clamps_negative_and_non_finite() {
        assert_eq!(line_total_minor(-1.0, 100), 0);
        assert_eq!(line_total_minor(f64::NAN, 100), 0);
        assert_eq!(line_total_minor(1.0, -50), 0);
    }
}

File: src/lib/money.ts
/**
 * All monetary amounts in the database are **integer minor units** (e.g. cents).
 * Use `bigint` for sums and products; use `formatMoneyMinor` only for **display** (uses `Intl`, not for accounting logic).
 */

/** Quantity scale matching Rust `domain::money::line_total_minor` (6 decimal places). */
const QTY_SCALE = 1_000_000n;

export function parseMinorInt(raw: string): number {
  const t = raw.trim().replace(/,/g, "");
  if (t === "" || t === "-") {
    return 0;
  }
  if (!/^-?\d+$/.test(t)) {
    throw new Error("Enter a whole number of minor units (e.g. cents).");
  }
  const n = Number(t);
  if (!Number.isFinite(n) || !Number.isInteger(n)) {
    throw new Error("Enter a whole number of minor units (e.g. cents).");
  }
  if (n > Number.MAX_SAFE_INTEGER || n < Number.MIN_SAFE_INTEGER) {
    throw new Error("Amount is too large for safe integer arithmetic.");
  }
  return n;
}

/** Coerce JSON / unknown values to a safe integer minor amount for display. */
export function asSafeMinor(value: unknown): number {
  if (typeof value === "bigint") {
    if (
      value > BigInt(Number.MAX_SAFE_INTEGER) ||
      value < BigInt(Number.MIN_SAFE_INTEGER)
    ) {
      throw new Error("Amount exceeds safe integer range.");
    }
    return Number(value);
  }
  const n = typeof value === "number" ? value : Number(value);
  if (!Number.isFinite(n) || !Number.isInteger(n)) {
    throw new Error("Amount must be an integer of minor units.");
  }
  if (!Number.isSafeInteger(n)) {
    throw new Error("Amount exceeds safe integer range.");
  }
  return n;
}

export function sumMinor(values: readonly number[]): number {
  let t = 0n;
  for (const v of values) {
    t += BigInt(Math.trunc(v));
  }
  if (
    t > BigInt(Number.MAX_SAFE_INTEGER) ||
    t < BigInt(Number.MIN_SAFE_INTEGER)
  ) {
    throw new Error("sum overflow");
  }
  return Number(t);
}

/**
 * Line total in minor units — mirrors Rust `line_total_minor`
 * (qty quantized to 6 dp, half-away-from-zero, clamp ≥ 0).
 */
export function lineTotalMinor(qty: number, unitMinor: number): number {
  if (!Number.isFinite(qty) || !Number.isFinite(unitMinor)) {
    return 0;
  }
  const qtyScaled = BigInt(Math.round(qty * Number(QTY_SCALE)));
  const product = qtyScaled * BigInt(Math.trunc(unitMinor));
  const half = QTY_SCALE / 2n;
  const rounded =
    product >= 0n ? (product + half) / QTY_SCALE : (product - half) / QTY_SCALE;
  if (rounded <= 0n) {
    return 0;
  }
  if (rounded > BigInt(Number.MAX_SAFE_INTEGER)) {
    return Number.MAX_SAFE_INTEGER;
  }
  return Number(rounded);
}

/** ISO 4217 fraction digits for a currency (fallback 2). */
export function currencyFractionDigits(
  currencyCode: string,
  locale: string = typeof navigator !== "undefined"
    ? navigator.language
    : "en-US",
): number {
  try {
    const digits = new Intl.NumberFormat(locale, {
      style: "currency",
      currency: currencyCode,
    }).resolvedOptions().maximumFractionDigits;
    return typeof digits === "number" ? digits : 2;
  } catch {
    return 2;
  }
}

/** Major-unit label for checks (e.g. USD → "Dollars"). */
export function currencyMajorLabel(
  currencyCode: string,
  locale: string = typeof navigator !== "undefined"
    ? navigator.language
    : "en-US",
): string {
  const code = currencyCode.trim().toUpperCase() || "USD";
  if (code === "USD") {
    return "Dollars";
  }
  try {
    return (
      new Intl.DisplayNames([locale], { type: "currency" }).of(code) ?? code
    );
  } catch {
    return code;
  }
}

export function formatMoneyMinor(
  minor: number | unknown,
  currencyCode: string = "USD",
  locale: string = typeof navigator !== "undefined"
    ? navigator.language
    : "en-US",
): string {
  const safe = asSafeMinor(minor);
  const digits = currencyFractionDigits(currencyCode, locale);
  const divisor = 10 ** digits;
  const major = digits === 0 ? safe : safe / divisor;
  return new Intl.NumberFormat(locale, {
    style: "currency",
    currency: currencyCode,
    minimumFractionDigits: digits,
    maximumFractionDigits: digits,
  }).format(major);
}
